server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java

```diff
@@ -153,10 +153,35 @@ public final XContentBuilder toXContent(XContentBuilder builder, Params params)
         return builder.endObject();
     }
 
+    /*
+     * By default metadata fields cannot be set through the document source and parse() method
+     * throws an exception. To enable a metadata field to parse the document source, this
+     * method must be overridden and doParse() should be called.
+     */
+    @Override
+    public void parse(ParseContext context) throws IOException {
+        throw new MapperParsingException("Field [" + name() + "] is a metadata field and cannot be added inside"
+            + " a document. Use the index API request parameters.");
+    }
+
+    /**
+     * Do the actual parse of the field by calling {@link FieldMapper#parse(ParseContext)}
+     */
+    protected void doParse(ParseContext context) throws IOException {
+        super.parse(context);
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
```

Comment:
Instead of overriding `parse` here, could we override `parseCreateField`?

```
@Override
protected void parseCreateField(ParseContext context) throws IOException {
    throw new MapperParsingException("Field [" + name() + "] is a metadata field and cannot be added inside"
        + " a document. Use the index API request parameters.");
}
```

Then we could do the following:
* For meta fields that cannot be specified in _source, move all the relevant logic out of `parseCreateField` and into `preParse` or `postParse` as appropriate.
* Remove the new `doParse` method, as it's no longer needed.

Reply:
Thanks for pointing this out. I had a hard time thinking a simple and clean way to refactor this part.

Initially, I tried the approach you suggested, but `FieldMapper#parse()` does all the exception handling by encapsulating all exceptions thrown by `parseCreateField()` offering a preview of the parsed value.

(See https://github.com/elastic/elasticsearch/blob/9a127adb4b1dd8e69ac5b6c08cc7251383562e9c/server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java#L249)

This means that the 
```
    throw new MapperParsingException("Field [" + name() + "] is a metadata field and cannot be added inside"
        + " a document. Use the index API request parameters.");
```

will be encapsulated in the following exception:

```
     throw new MapperParsingException("failed to parse field [{}] of type [{}] in document with id '{}'. " +
                "Preview of field's value: '{}'", e, fieldType().name(), fieldType().typeName(),
                context.sourceToParse().id(), valuePreview);
```

On the other hand, methods `preParse()` and `postParse()` delegate parsing to `parse()` for the same reason (to handle parsing exception). If we were to move all parsing in those two methods, we would have to replicate exception handling in those methods as well. 

Since all `FieldMapper#parse()` does is to delegate parsing to `parseCreateField()` and only do the exception handling, I decided to redirect this functionality in 'MetadataFieldMapper#doParse()` and call this method from `preParse()` and `postParse()`.

Reply:
To me that approach you tried is the cleanest way to go. My reasoning...

> This means that the ... will be encapsulated in the following exception:

This seems like an okay compromise to me, all the exception information is there so the user can determine the cause. It's a little confusing that we wrap the message, but this should be a rare situation.

> On the other hand, methods preParse() and postParse() delegate parsing to parse() for the same reason (to handle parsing exception). 

I think that exception handling code is only really helpful when dealing with incorrect user-supplied values. Since `preParse` and `postParse` work with internal data such as `context.sourceToParse().id()`, it would indicate a serious logic error. So I don't think we'd need the same general exception handling strategy where we create a nice user-facing message.